scripts/data/src/org/akvo/gae/remoteapi/ExtractImageGeotag.java

```diff
@@ -0,0 +1,292 @@
+
+        final Query qq = new Query("QuestionAnswerStore").setFilter(new Query.FilterPredicate("type", FilterOperator.EQUAL, "IMAGE"));
+        final PreparedQuery qpq = ds.prepare(qq);
+        List<Entity> questionsToFix = new ArrayList<Entity>();
+
+        for (Entity q : qpq.asIterable(FetchOptions.Builder.withChunkSize(500))) {
+            total++;
+            boolean forceSave = false;
+            Media media;
+            String v = (String) q.getProperty("value"); //Assume valueText was not used for this image
+            Date collected = (Date) q.getProperty("collectionDate");
+            if (collected.before(tooOld)) continue; //Debug! Concentrate on recent data.
+
+            System.out.printf(" Old IMAGE value '%s'\n", v);
+            if (v != null && !v.trim().equals("")) {
+                if (v.startsWith("{")) {
+                    json++;
+                } else {
+                    nonjson++;
+                    forceSave = true;//handle legacy values: convert them to JSON while we're here
+                    v = Paths.get(v).getFileName().toString(); //strip path, it is never used
+                }
+                String filename;
+                //Parse it
+                media = MediaResponse.parse(v);
+                if (media.getLocation() != null) { //Already known TODO check validity?
+                    //System.out.println("Location in " + q);
+                    tagged++;
+                    continue; //Skip
+                }
+                //also want to skip if location is present, but null, to avoid re-evaluation
+                if (v.matches("\"location\":null")) {
+                    System.out.printf("Null location in IMAGE %d: '%s'\n", q.getKey().getId(), v);
+                    tagged++;
+                    continue; //Skip
+                }
+
+
+            } else {
+                System.out.printf("#ERROR null or empty value for IMAGE %d: '%s'\n", q.getKey().getId(), v);
+                continue; //TODO convert this to JSON too?
+            }
+
+            //No location known; must read the file
+            Location loc = new Location();
+            Boolean tagFound = fetchLocationFromJpegInS3(media.getFilename(), loc);
+            if (tagFound != null || forceSave) {
+
+                if (tagFound == null) { // We cannot know (right now - file may arrive later)
+                    v = MediaResponse.formatWithoutGeotag(media);
+                } else { //We do know!
+                    if (tagFound.equals(Boolean.FALSE)) {
+                        loc = null; //There is no tag!
+                    }
+                    media.setLocation(loc);
+                    v = MediaResponse.formatWithGeotag(media);
+                }
+                System.out.printf(" New IMAGE value '%s'\n", v);
+                q.setProperty("value", v);
+                questionsToFix.add(q);
+                if (doIt && questionsToFix.size() >= 250) {
```

Comment:
I am not 100% sure but when you retrieve 500 and only save 250, would it maybe make sense to just retrieve them in batches of 250 in line 111?

Reply:
The batches are not equal. Only those values that make it past the filters end up in the toFix list.